### PR TITLE
feat(Apollo): Expose mutations to the gateway.

### PR DESCRIPTION
**Pull request by minhaj-shakeel**
Fixes GRAPHQL-968.
This PR does two things.
1 - It exposes the `mutations` to the apollo gateway which were earlier restricted. It exposes mutations for both `type definitions` as well as `type extensions`. Whereas `queries` for the `extended types` are `not exposed` as they are resolved through `entity resolver`.
2 - It adds support for add mutation on extended type having field with `ID` type as @key field. In this case the field is stored internally in the dgraph as `string` type so its value needs to be given as an argument inside the mutation.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)

### CODE REVIEW

**Review by abhimanyusinghgaur — APPROVED** [review]
Nice PR!
:lgtm: 

> Reviewed 7 of 7 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-:-MRsAmz6ArdbJMvX0Ktf:bcnbuob)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @pawanrawal)

**Review by pawanrawal — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: 6 of 9 files reviewed, 5 unresolved discussions (waiting on @abhimanyusinghgaur and @minhaj-shakeel)

___
*[graphql/e2e/common/query.go, line 455 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRsyQ8r1ERtQKKPGd_2:-MRsyQ8r1ERtQKKPGd_3:b-wk20) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/e2e/common/query.go#L455)):*
> ```Go
> func addMutationOnExtendedTypeWithIDasKeyField(t *testing.T) {
> 	addAstronautParams := &GraphQLParams{
> 		Query: `mutation addAstronaut($id1: ID!, $missionId1: String!, $id2: ID!, $missionId2: String! ) {
> ```

Do the mission and astronaut with the specified ids already exist in this service?
___
*[graphql/resolve/query_rewriter.go, line 263 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRsz63NET9rTsPc9J-S:-MRsz63NET9rTsPc9J-T:b-ol15mk) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/resolve/query_rewriter.go#L263)):*
> ```Go
> 	//	}
> 
> 	// If key field is of ID type but it is an external field,
> ```

This comment looks like it belongs in the else block. 

So basically if it's not external, then we query using the uid otherwise we treat it as a string and query using `eq` function.
___
*[graphql/schema/gqlschema.go, line 993 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRszZWF7mN3c_1ji_TA:-MRszZWF7mN3c_1ji_TB:baxxoym) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/gqlschema.go#L993)):*
> ```Go
> 		addFieldFilters(sch, defn, apolloServiceQuery)
> 		addAggregationResultType(sch, defn)
> 		// Don't expose queries for the @extends type to the gateway
> ```

What about all the other things like filters and such which the query uses. Should they also not be exposed?
___
*[graphql/schema/wrappers.go, line 511 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt--7z47yD6yaUpLG1:-MRt--7z47yD6yaUpLG2:b79zxsb) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/wrappers.go#L511)):*
> ```Go
> 			// If key field is of ID type but it is an external field,
> 			// then it is stored in Dgraph as string type with Hash index.
> 			// So we need the predicate mapping in this case.
> ```

Can we add a dgraph schema gen test for key field which is an ID but is an external field that shows the type for it in Dgraph schema?
___
*[graphql/schema/testdata/apolloservice/output/extended-types.graphql, line 357 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt-Wgw04eSjTAfvbVC:-MRt-Wgw04eSjTAfvbVD:b-8jqv0s) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/testdata/apolloservice/output/extended-types.graphql#L357)):*
> ```GraphQL
> }
> 
> input AstronautFilter {
> ```

If the astronaut query isn't generated then why do we have AstronautFilter and ORder.

**Review by pawanrawal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: 6 of 9 files reviewed, 2 unresolved discussions (waiting on @abhimanyusinghgaur and @minhaj-shakeel)

___
*[graphql/e2e/common/query.go, line 466 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt6Qrv9IY0vDn0HfxR:-MRt6Qrw5iWko7Z4xZxM:b-xurx22) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/e2e/common/query.go#L466)):*
> ```Go
> 			}
> 		}`,
> 		Variables: map[string]interface{}{
> ```

Can you add a test where you add a new mission and link to an astronaut that doesn't exist yet? Does that return an error or work properly?

**Review by pawanrawal — COMMENTED** [review]
> Reviewed 6 of 7 files at r1, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: all files reviewed, 2 unresolved discussions (waiting on @minhaj-shakeel)

**Review by minhaj-shakeel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: 6 of 12 files reviewed, 2 unresolved discussions (waiting on @abhimanyusinghgaur and @pawanrawal)

___
*[graphql/e2e/common/query.go, line 466 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt6Qrv9IY0vDn0HfxR:-MS1P4toE9YdHZOhzy40:b-wpzbsd) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/e2e/common/query.go#L466)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Can you add a test where you add a new mission and link to an astronaut that doesn't exist yet? Does that return an error or work properly?
</blockquote></details>

Done that. It creates a new object or links the object if the object with the same `xid` already exists.
___
*[graphql/resolve/query_rewriter.go, line 263 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRsz63NET9rTsPc9J-S:-MS1PIF30FhHXixFiCj-:bbhqgb4) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/resolve/query_rewriter.go#L263)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

This comment looks like it belongs in the else block. 

So basically if it's not external, then we query using the uid otherwise we treat it as a string and query using `eq` function.
</blockquote></details>

Yeah, shifted the comment and modified it.

**Review by minhaj-shakeel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: 6 of 12 files reviewed, 2 unresolved discussions (waiting on @abhimanyusinghgaur and @pawanrawal)

___
*[graphql/e2e/common/query.go, line 455 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRsyQ8r1ERtQKKPGd_2:-MS1QFpXDPpOxJsn0-w8:bfu7iwe) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/e2e/common/query.go#L455)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Do the mission and astronaut with the specified ids already exist in this service?
</blockquote></details>

No, both don't exist. If the mission exists, it will throw an error and if the Astronaut exists then it will just link the previous object the newly created mission object.
___
*[graphql/schema/gqlschema.go, line 993 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRszZWF7mN3c_1ji_TA:-MS1QgB9DtjaOHgo1eGG:b-gk1j0u) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/gqlschema.go#L993)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

What about all the other things like filters and such which the query uses. Should they also not be exposed?
</blockquote></details>

These are being exposed. See the `schemagen/apolloservice` tests.
___
*[graphql/schema/wrappers.go, line 511 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt--7z47yD6yaUpLG1:-MS1QmqbCxQo9EGEk0_u:blodwiv) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/wrappers.go#L511)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Can we add a dgraph schema gen test for key field which is an ID but is an external field that shows the type for it in Dgraph schema?
</blockquote></details>

Those are already added in the `federation` branch.
___
*[graphql/schema/testdata/apolloservice/output/extended-types.graphql, line 357 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7358#-MRt-Wgw04eSjTAfvbVC:-MS1QsnI3c03qMj3aRSo:bvaw3oo) ([raw file](https://github.com/dgraph-io/dgraph/blob/b2a2adc2af1f1915732267bef6acd8a1f19e9630/graphql/schema/testdata/apolloservice/output/extended-types.graphql#L357)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

If the astronaut query isn't generated then why do we have AstronautFilter and ORder.
</blockquote></details>

For `AddAstronautPayload`.

**Review by pawanrawal — APPROVED** [review]
> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7358)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @minhaj-shakeel)
